Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp

```diff
@@ -0,0 +1,349 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "AudioDecoderCocoa.h"
+
+#if ENABLE(WEB_CODECS) && USE(AVFOUNDATION)
+
+#include "AudioSampleBufferConverter.h"
+#include "AudioSampleFormat.h"
+#include "CAAudioStreamDescription.h"
+#include "MediaSampleAVFObjC.h"
+#include "MediaUtilities.h"
+#include "PlatformRawAudioData.h"
+#include "SharedBuffer.h"
+#include <CoreAudio/CoreAudioTypes.h>
+#include <wtf/NeverDestroyed.h>
+#include <wtf/TZoneMallocInlines.h>
+#include <wtf/UniqueRef.h>
+#include <wtf/WorkQueue.h>
+#include <wtf/text/MakeString.h>
+
+#include <pal/cf/AudioToolboxSoftLink.h>
+#include <pal/cf/CoreMediaSoftLink.h>
+
+namespace WebCore {
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(AudioDecoderCocoa);
+
+static WorkQueue& queueSingleton()
```

Comment:
Do we need this queue since `AudioSampleBufferConverter` is doing the work in a queue itself?

Reply:
On further thinking I think that at this stage, yes we do.
The AudioSampleBufferConverter uses both NativePromise and a callback on its queue.

To handle the NativePromise we need to know where the result should be dispatched. Having a dedicated queue is the simplest approach.

Otherwise, we would need to pass the WorkerThread to the AudioDecoderCocoa and use enqueueWhenSettled on the context.

2) the callback as it is managed only take a raw pointer, so making sure that pointer is valid to the end makes it not really suitable with a WorkerThread which could be deleted at any given time. We would need to dispatch a task regardless on the converter to finish it.

So I can look at this in a follow up task, but considering all {Audio|Video){Encoder|Decoder) follow the exact same patterns on using their own queue, I'm not sure why here we would do anything different particularly when it would make the code harder, not easier IMHO.